os/arch/arm/src/amebad/amebad_reboot_reason.c

```diff
@@ -70,12 +75,38 @@ void up_reboot_reason_init(void)
 
 reboot_reason_code_t up_reboot_reason_read(void)
 {
-	return 0;
+	u32 backup_reg = 0;
+	u32 boot_reason = 0;
+	u32 temp_bootcfg = 0;
+
+	/* Read the same backup register for the boot reason */
+	backup_reg = BKUP_Read(BKUP_REG1);
+
+	/* Read AmebaD Boot Reason, WDT and HW reset supported */
+	boot_reason = BOOT_Reason();
+
+	/* HW reset */
+	if (boot_reason == 0) {
+		temp_bootcfg |= UPPER_REBOOT_SYSTEM_HW_RESET;
+	}
+
+	/* KM4 or KM0 WDT reset */
+	else if ((boot_reason & BIT_BOOT_KM4WDG_RESET_HAPPEN) || (boot_reason & BIT_BOOT_WDG_RESET_HAPPEN)) {
+		temp_bootcfg |= UPPER_REBOOT_SYSTEM_WATCHDOG;
+	}
+
+	temp_bootcfg |= backup_reg;
```

Comment:
In this case, is `temp_bootcfg` matched with `reboot_reason_code_t` in `tinyara/reboot_reason.h`?

Reply:
Yes

Reply:
Ok. Thanks.

Reply:
I see that in reboot_reason.h, REBOOT_SYSTEM_HW_RESET is defined to be BIT(4).

Reply:
Oh sorry, I mean for `REBOOT_SYSTEM_WATCHDOG` case.

Reply:
Some boot reason you provided in reboot_reason.h is not supported by us. We thought that you want to write your own boot reason and read, that's why we provide BKUP_REG to achieve it. The common cases we are both handling are:
BIT_BOOT_KM4WDG_RESET_HAPPEN BIT_BOOT_WDG_RESET_HAPPEN - REBOOT_SYSTEM_WATCHDOG
HW RESET - REBOOT_SYSTEM_HW_RESET

That is why we have the if-else if in up_reboot_reason_read.
Thanks.

Reply:
Okay. Thanks.

Reply:
In our usecase, we don't take combined two more reason. When we call `up_reboot_reason_read`, it should return only one value which is defined in tinyara/reboot_reason.h. Because reboot reason must be one.
The `sys_reset` Realtek provided is working with watchdog. The system will be reset by watchdog, but it's not real reboot reason. Because of some other reason, something calls sys_reset to reset. At that case, we will write the reason through `up_reboot_reason_write`.
Application should be the the same regardless of board so that please make this API to return just one reason.

Reply:
Please find our API prototype. `reboot_reason_code_t up_reboot_reason_read(void)`. It's not u32 for return.

Reply:
Does that mean we don't need to deal with the BIT_BOOT_KM4WDG_RESET_HAPPEH and the BIT_BOOT_WDG_RESET_HAPPEN and just leave it to you to take care of? How about HW reset? Do we set bit 4 or return 0 directly?

Reply:
For HW Reset, 4. (Not BIT(4))
```
REBOOT_SYSTEM_HW_RESET		= 4,	/* HW power reset */
```
For Watchdog time out, 0. (Not Bit(0) to set LSB)
```
REBOOT_SYSTEM_WATCHDOG		= 0,	/* watchdog timeout */  
```

Reply:
I think when `backup_reg = BKUP_Read(BKUP_REG1);` has some value, you can return it directly without checking `boot_reason = BOOT_Reason();`. When backup does not have, then you can check boot reason and then can return.

But we don't clear it all the time, simply reading backup and then boot reason could be return old reason.
How about reading the reason using the init function and saving it in static global variable? And read function returns just static global. Then we can prevent reading old reason.

Reply:
Thank you for your suggestions, I will try it out.